File: java/gov/nasa/gsfc/gmsec/api/field/CharField.java
/*
 * Copyright 2007-2021 United States Government as represented by the
 * Administrator of The National Aeronautics and Space Administration.
 * No copyright is claimed in the United States under Title 17, U.S. Code.
 * All Rights Reserved.
 */


/**
 * @file CharField.java
 */

package gov.nasa.gsfc.gmsec.api.field;

import gov.nasa.gsfc.gmsec.api.jni.field.JNICharField;


/**
 * Specialized Field class that can be used to retain a single character value.
 */
public class CharField extends Field
{
	/**
	 * This constructor is for internal GMSEC API use only.
	 * @param field A JNICharField object.
	 */
	public CharField(JNICharField field)
	{
		setInternal(field);
	}


	/**
	 * Default constructor.
	 *
	 * @param name Name of the field.
	 * @param data Character data to associate with the field.
	 *
	 * @throws IllegalArgumentException Thrown if the field name is null, or contains an empty string.
	 */
	public CharField(String name, char data) throws IllegalArgumentException
	{
		validateName(name);

		setInternal(new JNICharField(name, data));
	}


	/**
	 * Copy constructor.
	 *
	 * @param other The other CharField object to copy.
	 *
	 * @throws IllegalArgumentException Thrown if the given CharField is null.
	 */
	public CharField(CharField other) throws IllegalArgumentException
	{
		if (other == null)
		{
			throw new IllegalArgumentException("CharField object is null");
		}

		setInternal(new JNICharField((JNICharField) Field.getInternal(other)));
	}


	/**
	 * Returns the data stored within the CharField.
	 *
	 * @return A char value.
	 */
	public char getValue()
	{
		JNICharField jCharField = (JNICharField) getInternal();

		return jCharField.getValue();
	}
}

File: framework/src/gmsec4/c/field/char_field.cpp
/*
 * Copyright 2007-2021 United States Government as represented by the
 * Administrator of The National Aeronautics and Space Administration.
 * No copyright is claimed in the United States under Title 17, U.S. Code.
 * All Rights Reserved.
 */



/** @file char_field.cpp
 *
 *  @brief This file contains functions for the management of a CharField object.
 */


#include <gmsec4/c/field/char_field.h>

#include <gmsec4/field/CharField.h>

#include <gmsec4/Exception.h>
#include <gmsec4/Status.h>


using namespace gmsec::api;


GMSEC_Field CALL_TYPE charFieldCreate(const char* fieldName, GMSEC_CHAR value, GMSEC_Status status)
{
	CharField* field = NULL;
	Status     result;

	try
	{
		field = new CharField(fieldName, value);
	}
	catch (Exception& e)
	{
		result = Status(e);
	}

	if (status)
	{
		*(reinterpret_cast<Status*>(status)) = result;
	}

	return field;
}


GMSEC_CHAR CALL_TYPE charFieldGetValue(const GMSEC_Field field, GMSEC_Status status)
{
	const Field*     f         = reinterpret_cast<const Field*>(field);
	const CharField* charField = dynamic_cast<const CharField*>(f);
	GMSEC_CHAR       value     = 0;
	Status           result;

	if (charField)
	{
		value = charField->getValue();
	}
	else
	{
		result = Status(FIELD_ERROR, INVALID_FIELD, "Field object is not a CharField");
	}

	if (status)
	{
		*(reinterpret_cast<Status*>(status)) = result;
	}

	return value;
}
